projects/packages/jetpack-mu-wpcom/src/features/calypso-locale-sync/sync-wp-admin-to-calypso.php

```diff
@@ -27,32 +30,42 @@ function update_calypso_locale( $locale ) {
 
 /**
  * Sync locale updated via /wp-admin/profile.php to Calypso.
- *
- * @param array   $meta   Meta values and keys for the user.
- * @param WP_User $user   User object.
- * @param boolean $update Whether the user is being updated rather than created.
  */
-function sync_wp_admin_locale_to_calypso( $meta, $user, $update ) {
-	$locale            = $meta['locale'];
-	$old_locale        = get_user_locale( $user );
-	$is_user_connected = ( new Connection_Manager( 'jetpack' ) )->is_user_connected();
+function sync_wp_admin_locale_on_profile_update() {
+	$locale = null;
+
+	if ( isset( $_POST['locale'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
+		$locale = sanitize_text_field( wp_unslash( $_POST['locale'] ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+	}
+
+	if ( '' === $locale ) {
+		$locale = 'en';
+	}
+
+	$user_id = get_current_user_id();
 
-	if ( ! $update || $old_locale === $locale || ! $is_user_connected ) {
+	if ( ! $user_id || empty( $locale ) ) {
+		return;
+	}
+
+	$user               = get_userdata( $user_id );
+	$old_locale         = get_user_locale( $user );
+	$connection_manager = new Connection_Manager( 'jetpack' );
+	$is_user_connected  = $connection_manager->is_user_connected();
+
+	if ( $old_locale === $locale || ! $is_user_connected ) {
```

Comment:
According to the post-mortem post, "_there should only have been one request “allowed” per HTTP request_", so let's flag on a static variable whether we're already syncing the locale and bail if that's the case.

Reply:
Additionally, we could also follow the "_further restrict this to only running in a logged-in wp-admin context or wp-json-specific context, not in a front-page-unlogged-in context_" advice, and make sure that this only runs if `is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST )`

Reply:
Done in https://github.com/Automattic/jetpack/pull/37750/commits/07326acc9a25bd66039684d8f412b9b0976b7017 and https://github.com/Automattic/jetpack/pull/37750/commits/13ae7cb41a0b4272125bca41dfa007d6d3791b5a